test/net/sourceforge/kolmafia/textui/javascript/AshInteropTest.java

```diff
@@ -0,0 +1,34 @@
+package net.sourceforge.kolmafia.textui.javascript;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+import net.sourceforge.kolmafia.KoLmafia;
+import net.sourceforge.kolmafia.session.ContactManager;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+
+public class AshInteropTest {
+  @BeforeAll
+  static void clearStatus() {
+    KoLmafia.lastMessage = "";
+  }
+
+  @Test
+  void getPlayerIdReturnsInt() {
+    // Note that we won't do any actual lookups, since ChatManager.isLiterate is false.
+    ContactManager.registerPlayerId("heeheehee", "354981");
+
+    var js = new JavascriptRuntime("getPlayerId('heeheehee')");
+    assertEquals("", KoLmafia.lastMessage);
+    assertEquals("354981", js.execute(null, null, true).toString());
+  }
+
+  @Test
+  void getPlayerIdInvertsGetPlayerName() {
+    ContactManager.registerPlayerId("heeheehee", "354981");
+
+    var js = new JavascriptRuntime("getPlayerName(parseInt(getPlayerId('heeheehee')))");
+    assertEquals("", KoLmafia.lastMessage);
```

Comment:
let's simplify this.  You're already sure you're getting an Int from getPlayerIdReturnsInt() above.  Don't chain it here, so we don't test three functions.

Reply:
Actually, that test name lies. We're getting a string from getPlayerId.

Reply:
Either way, simplified to just testing a single function in each case, and replaced comparisons against v.toString() with Value comparisons. It's not much better, since equals does coercions, but at least it won't coerce numbers to strings.

Maybe I should instead be using rawValue... although that doesn't perform any type checking either.

Reply:
Thing is, _both_ test cases were failing under Windows. So even the simple one that was just doing getPlayerId('heeheehee') was returning null.

Reply:
```
AshInteropTest > getPlayerIdReturnsString() FAILED

    org.opentest4j.AssertionFailedError at AshInteropTest.java:24


AshInteropTest > getPlayerIdNameCoercesArgumentToInt() FAILED

    org.opentest4j.AssertionFailedError at AshInteropTest.java:32
```
the first one was passing before, IIRC.

the second one is weird, isn't it?  What's in LastMessage on that line that's breaking the 2nd test.

Reply:
> the first one was passing before, IIRC.

We can actually see historical executions in the Actions tab.

https://github.com/kolmafia/kolmafia/actions?query=branch%3Ajs-coercion

https://github.com/kolmafia/kolmafia/runs/3826425362

```
AshInteropTest > getPlayerIdReturnsInt() FAILED
    java.lang.NullPointerException at AshInteropTest.java:15

AshInteropTest > getPlayerIdInvertsGetPlayerName() FAILED
    java.lang.NullPointerException at AshInteropTest.java:23
```

I have a second branch whose sole purpose is to publish reports on commit (since cross-repo reports don't get published due to security concerns) so I can see more report details.

Results of latest check: https://github.com/heeheehee-kolmafia/kolmafia/runs/3834392716

Reply:
My git-fu is not advanced enough to work with this branch.  I think it is because I can make my tools work for branches of kolmafia but not branches of heeheehee-kolmafia.  I did clone h3-k but then could not get the branch.  In any event, I extracted the one test file and put it in my own local branch.  There's an educational opportunity :-) 

I know the point of this PR is that JS doesn't have strong typing and what it does have needs to behave with Ash.  The only test failure I had was in what had been the last test and that was because 354981 was not equal to 354981.0 .  I tracked that down to js_parseInt which took 354981 as input and returned 354981.0  My conclusion was that either js_parseInt is buggy or we are using it inappropriately.  

I used the debugger a lot and unchained some things so that I could inspect an intermediate object or two.  Everything I looked at but the one conversion behaved as expected.  Breakpoints in the Ash functions showed the expected inputs and outputs.  The new test I added worked.  Basically we are composing two functions where f(x) = y, g(y) = x and thus f(g(y)) = g(f(x)) and my new test made sure both "singletons" were tested separately.

I was thinking I was looking at a test that had failed on Windows but passed in other environment.  It does sound like there are various failures that may not be the same across OSes.

I've got a couple of (time) windows later today and will try and see if I can get the null pointer and other reported errors.

Reply:
Missed there were changes there...  Will fetch